libcaf_test/caf/test/fixture/flow.hpp

```diff
@@ -144,6 +144,81 @@ class CAF_TEST_EXPORT flow {
     caf::flow::coordinator* parent_;
   };
 
+  /// A subscription implementation without internal logic.
+  class passive_subscription_impl final
+    : public caf::flow::subscription::impl_base {
+  public:
+    explicit passive_subscription_impl(caf::flow::coordinator* parent)
+      : parent_(parent) {
+      // nop
+    }
+
+    /// Incremented by `request`.
+    size_t demand = 0;
+
+    /// Flipped by `dispose`.
+    bool disposed_flag = false;
+
+    caf::flow::coordinator* parent() const noexcept override;
+
+    void request(size_t n) override;
+
+    bool disposed() const noexcept override;
+
+  private:
+    void do_dispose(bool from_external) override;
+
+    caf::flow::coordinator* parent_;
+  };
+
+  template <class T>
+  class canceling_observer : public caf::flow::observer_impl_base<T> {
+  public:
+    explicit canceling_observer(caf::flow::coordinator* parent,
+                                bool accept_first)
+      : accept_next(accept_first), parent_(parent) {
+      // nop
+    }
+
+    caf::flow::coordinator* parent() const noexcept override {
+      return parent_;
+    }
+
+    void on_next(const T&) override {
+      ++on_next_calls;
+      sub.cancel();
+    }
+
+    void on_error(const error&) override {
+      ++on_error_calls;
+      sub.release_later();
+    }
+
+    void on_complete() override {
+      ++on_complete_calls;
+      sub.release_later();
+    }
+
+    void on_subscribe(caf::flow::subscription sub) override {
+      if (accept_next) {
+        accept_next = false;
+        sub.request(128);
+        this->sub = std::move(sub);
+        return;
+      }
+      sub.cancel();
+    }
+
+    int on_next_calls = 0;
+    int on_error_calls = 0;
+    int on_complete_calls = 0;
+    bool accept_next = false;
```

Comment:
I think this would be a better name:

```suggestion
    bool accept_subscription = false;
```